File: db/migrate/0003_create_posts.rb
# -*- coding: utf-8 -*-

class CreatePosts < ActiveRecord::Migration
  def change
    create_table(:posts) do |t|
      ## Database authenticatable
      t.string :slug, null: false
      t.string :guid, null: false
      t.boolean :visible, null: false

      t.integer :blog_id, null: false
      t.integer :author_id, null: false

      t.string :subject, null: false
      t.text :excerpt
      t.text :content, null: false
      t.string :format, null: false, default: 'html'

      t.json :attributes, null: false, default: '{}'

      t.timestamp :published, null: false

      t.timestamps
    end


    add_index :posts, :guid, :unique => true
    add_index :posts, :slug, :unique => true

    add_index :posts, :blog_id

    reversible do |dir|
      dir.up do
        execute "ALTER TABLE posts ADD CONSTRAINT blog_id_fk FOREIGN KEY (blog_id) REFERENCES blogs (id) ON DELETE CASCADE ON UPDATE CASCADE;"
        execute "ALTER TABLE posts ADD CONSTRAINT author_id_fk FOREIGN KEY (author_id) REFERENCES authors (id) ON DELETE CASCADE ON UPDATE CASCADE;"
      end

      dir.down do
        execute "ALTER TABLE posts DROP CONSTRAINT blog_id_fk"
        execute "ALTER TABLE posts DROP CONSTRAINT author_id_fk"
      end
    end
  end
end

# eof

File: priv/repo/migrations/0003_create_posts.exs
defmodule Burp.Repo.Migrations.CreatePosts do
  use Ecto.Migration

  def change do
    create table(:posts) do
      add :slug, :string, null: false
      add :guid, :string, null: false
      add :visible, :boolean, default: false, null: false

      add :subject, :string, null: false
      add :excerpt, :text
      add :content, :text, null: false
      add :format, :string, null: false, default: "html"
      add :attrs, :map, null: false, default: %{}
      add :published_at, :naive_datetime, null: false

      add :blog_id, references(:blogs, on_delete: :delete_all, on_update: :update_all)
      add :author_id, references(:authors, on_delete: :delete_all, on_update: :update_all)

      timestamps()
    end

    create unique_index(:posts, [:slug])
    create unique_index(:posts, [:guid])

    create index(:posts, [:blog_id])
    create index(:posts, [:author_id])
  end
end
